solr/core/src/java/org/apache/solr/search/facet/FacetIntervalProcessor.java

```diff
@@ -0,0 +1,154 @@
+    //@todo apoorv handle exception, sort orders
+    List<Map<String, String>> intervals = (List<Map<String, String>>) input;
+    List<Range> ranges = new ArrayList<>();
+    for (Map<String, String> interval : intervals) {
+      String key = interval.get("key");
+      if (interval == null) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "empty facet interval");
+      }
+      String intervalStr = interval.get("value").trim();
+      if (intervalStr.length() == 0) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "empty facet interval");
+      }
+      Boolean startOpen = false, endOpen = false;
+      Comparable start = null, end = null;
+      if (intervalStr.charAt(0) == '(') {
+        startOpen = true;
+      } else if (intervalStr.charAt(0) == '[') {
+        startOpen = false;
+      } else {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Invalid start character " + intervalStr.charAt(0) + " in facet interval " + intervalStr);
+      }
+
+      final int lastNdx = intervalStr.length() - 1;
+      if (intervalStr.charAt(lastNdx) == ')') {
+        endOpen = true;
+      } else if (intervalStr.charAt(lastNdx) == ']') {
+        endOpen = false;
+      } else {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Invalid end character " + intervalStr.charAt(0) + " in facet interval " + intervalStr);
+      }
+
+      StringBuilder startStr = new StringBuilder(lastNdx);
+      int i = unescape(intervalStr, 1, lastNdx, startStr);
+      if (i == lastNdx) {
+        if (intervalStr.charAt(lastNdx - 1) == ',') {
+          throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Empty interval limit");
+        }
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Missing unescaped comma separating interval ends in " + intervalStr);
+      }
+      try {
+        start = calc.getValue(startStr.toString());
+      } catch (SolrException e) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, String.format(Locale.ROOT, "Invalid start interval for key '%s': %s", key, e.getMessage()), e);
+      }
+      StringBuilder endStr = new StringBuilder(lastNdx);
+      i = unescape(intervalStr, i, lastNdx, endStr);
+      if (i != lastNdx) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Extra unescaped comma at index " + i + " in interval " + intervalStr);
+      }
+      try {
+        end = calc.getValue(endStr.toString());
+      } catch (SolrException e) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, String.format(Locale.ROOT, "Invalid end interval for key '%s': %s", key, e.getMessage()), e);
+      }
+      if (start != null && end != null && start.compareTo(end) > 0) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Start is higher than end in interval for key: " + key);
+      }
+      if (interval.get("key") == null) key = startStr.toString();
```

Comment:
why not full interval range expression?

Reply:
This is what `key` defaults to in the older apis, keeping whole interval range expression would make more sense, will change